Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -14,173 +14,222 @@ import Data.Semigroup
+smoothLogSieveM sievingIntervalM base a b c m =
+  forM_ base $ \(prime, roots) -> case someNatVal (intToNatural (unPrime prime)) of
+    SomeNat (Proxy :: Proxy prime) -> do
+      let
+        startingIndices = case MW.invertMod (fromInteger a :: MW.Mod prime) of
+          Just inverseOfA -> map (\root -> (wordToInt . MW.unMod) (fromIntegral m + fromInteger (- b + root) * inverseOfA :: MW.Mod prime)) roots
+          Nothing         -> case MW.invertMod (fromInteger (2 * b) :: MW.Mod prime) of
+            -- Temporary fix. If power > 2, something better has to be found.
+            Just inverseOf2B -> [(wordToInt . MW.unMod) (fromIntegral m - fromInteger c * inverseOf2B :: MW.Mod prime)]
+            -- For this to be true 2 cannot be a factor of a.
+            -- Better to ouput @unPrime prime@ as a factor.
+            Nothing          -> error ("Found an illegal factor: " ++ show prime)
+      forM_ startingIndices $ \startingIndex -> do
+        let change y = y - ((intLog2 . unPrime) prime)
+        forM_ [startingIndex, startingIndex + (unPrime prime)..(2 * m)] $ \entry ->
+          -- let change (y, im) = (if y `mod` (intToInteger . unPrime) prime == 0 then y `div` (intToInteger . unPrime) prime else error ("Sieve fail: " ++ show (prime, power, entry)), I.insert (unPrime prime) power im)
+          MV.modify sievingIntervalM change entry
 
 -- This algorithm filters @sievingIntervalF@ for smooth numbers. It returns
 -- the smooth numbers together with their @index@ and @startingPoint@.
 -- This is needed in order to later compute @firstSquare@.
-findSmoothNumbers :: Integer -> V.Vector (Integer, SignedPrimeIntSet) -> V.Vector (Integer, SignedPrimeIntSet)
-findSmoothNumbers startingPoint = V.imapMaybe selectSmooth
+findLogSmoothNumbers :: Int -> Int -> Integer -> Integer -> V.Vector (Integer, Int) -> V.Vector (Integer, I.IntMap Int)
+findLogSmoothNumbers t m a b = V.imapMaybe selectSmooth
   where
-    selectSmooth index (residue, factorisation)
-      | residue == 1 = Just (intToInteger index + startingPoint, factorisation)
-      | otherwise    = Nothing
-
--- Find all primes, which appear only once in the input list.
-appearsOnlyOnce :: [PS.PrimeIntSet] -> PS.PrimeIntSet
-appearsOnlyOnce = fst . L.foldl' go (mempty, mempty)
-  where
-    go (onlyOnce, atLeastOnce) x =
-      ((onlyOnce PS.\\ PS.unPrimeIntSet x) <> (x PS.\\ PS.unPrimeIntSet atLeastOnce), atLeastOnce <> x)
+    -- Adjust fudge factor to taste
+    selectSmooth index (value, logResidue) = case logResidue < 18 of
+      True
+        -- Cannot use factorisation instead of listFactorisation since its rightmost value need not be an int.
+        | null listFactorisation || (fst . last) listFactorisation <= intToInteger t -> Just (a * intToInteger (index - m) + b, factorisation)
+        | otherwise                                                                  -> Nothing
+      False                                                                          -> Nothing
```

Comment:
Do not take alignment to such extremes. Better put `Just` / `Nothing` on a next line.